packages/eth-providers/src/base-provider.ts

```diff
@@ -1774,10 +1775,22 @@ export abstract class BaseProvider extends AbstractProvider {
     }
 
     const filter = await this._sanitizeRawFilter(rawFilter);
-    const subqlLogs = await this.subql.getFilteredLogs(filter); // only filtered by blockNumber and address
-    const filteredLogs = subqlLogs.filter((log) => filterLogByTopics(log, filter.topics));
 
-    return filteredLogs.map((log) => this.formatter.filterLog(log));
+    // make sure subql already indexed all target blocks, up until the latest finalized block
+    const upperBoundry = await this.finalizedBlockNumber;
+    const targetBlock = filter.toBlock <= upperBoundry
+      ? filter.toBlock
+      : upperBoundry;
+    let lastProcessedHeight = await this.subql.getLastProcessedHeight();
+    while (lastProcessedHeight < targetBlock) {
+      await sleep(1000);
+      lastProcessedHeight = await this.subql.getLastProcessedHeight();
+    }
```

Comment:
If if `toBlock` is very large, will this hang?

Reply:
so we have this https://github.com/AcalaNetwork/bodhi.js/pull/828/files#diff-1dc05f5efb9f973bee2a8da85d8839f8f96514217f7336e218b5c2535bd20baeR1780-R1783

Reply:
if toblock is larger than latest finalized block, we will use latest finalized block